src/parseable.rs

```diff
@@ -0,0 +1,232 @@
+/*
+ * Parseable Server (C) 2022 - 2024 Parseable, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ *
+ */
+
+use std::{collections::HashMap, path::PathBuf, sync::Arc};
+
+use arrow_schema::{Field, Schema};
+use bytes::Bytes;
+use clap::{error::ErrorKind, Parser};
+use once_cell::sync::Lazy;
+
+use crate::{
+    cli::{Cli, Options, StorageOptions},
+    handlers::http::logstream::error::StreamError,
+    metadata::{error::stream_info::LoadError, LogStreamMetadata, StreamInfo, LOCK_EXPECT},
+    option::Mode,
+    storage::{
+        object_storage::parseable_json_path, ObjectStorageError, ObjectStorageProvider,
+        ObjectStoreFormat,
+    },
+};
+
+pub const JOIN_COMMUNITY: &str =
+    "Join us on Parseable Slack community for questions : https://logg.ing/community";
+
+/// Shared state of the Parseable server.
+pub static PARSEABLE: Lazy<Parseable> = Lazy::new(|| match Cli::parse().storage {
+    StorageOptions::Local(args) => {
+        if args.options.local_staging_path == args.storage.root {
+            clap::Error::raw(
+                ErrorKind::ValueValidation,
+                "Cannot use same path for storage and staging",
+            )
+            .exit();
+        }
+
+        if args.options.hot_tier_storage_path.is_some() {
+            clap::Error::raw(
+                ErrorKind::ValueValidation,
+                "Cannot use hot tier with local-store subcommand.",
+            )
+            .exit();
+        }
+
+        Parseable::new(args.options, Arc::new(args.storage), "drive")
+    }
+    StorageOptions::S3(args) => Parseable::new(args.options, Arc::new(args.storage), "s3"),
+    StorageOptions::Blob(args) => {
+        Parseable::new(args.options, Arc::new(args.storage), "blob-store")
+    }
+});
+
+/// All state related to parseable, in one place.
+pub struct Parseable {
+    /// Configuration variables for parseable
+    pub options: Options,
+    /// Storage engine backing parseable
+    pub storage: Arc<dyn ObjectStorageProvider>,
+    /// Either "drive"/"s3"/"blob-store"
+    pub storage_name: &'static str,
```

Comment:
`storage_name` should be a part of the `storage`, why separate it?

Reply:
There is so much to organize, all these patterns are part of legacy code

Reply:
But they were separate things, now they are being refactored and brought in the single struct.

Reply:
Yes, the idea with creating atomic PRs is to ensure that we can track and undo changes that have side effects elsewhere, this PR's main aim was to package stuff related to server state in one place so that we can later perform simplifications along the ingestion path. I am not giving too much thought to anything outside of that path and this didn't fall into my radar as a big enough issue 😅